services/gossip/adapter/direct_transport_test.go

```diff
@@ -0,0 +1,278 @@
+package adapter
+
+import (
+	"context"
+	"fmt"
+	"github.com/orbs-network/orbs-network-go/test"
+	"github.com/orbs-network/orbs-spec/types/go/primitives"
+	"github.com/orbs-network/orbs-spec/types/go/protocol/gossipmessages"
+	"github.com/stretchr/testify/require"
+	"net"
+	"testing"
+)
+
+func TestDirectIncoming_ConnectionsAreListenedToWhileContextIsLive(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	h := newDirectHarness().start(ctx)
+
+	connection, err := net.Dial("tcp", fmt.Sprintf("127.0.0.1:%d", h.myPort))
+	defer connection.Close()
+	require.NoError(t, err, "test peer should be able connect to local transport")
+
+	cancel()
+
+	buffer := []byte{0}
+	read, err := connection.Read(buffer)
+	require.Equal(t, 0, read, "test peer should disconnect from local transport without reading anything")
+	require.Error(t, err, "test peer should disconnect from local transport")
+
+	_, err = net.Dial("tcp", fmt.Sprintf("127.0.0.1:%d", h.myPort))
+	require.Error(t, err, "test peer should be able to connect to local transport")
+}
+
+func TestDirectOutgoing_ConnectionsToAllPeersOnInitWhileContextIsLive(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	h := newDirectHarnessWithConnectedPeers(t, ctx)
+	defer h.cleanupConnectedPeers()
+
+	cancel()
+
+	for i := 0; i < networkSize-1; i++ {
+		buffer := []byte{0}
+		read, err := h.peersListenersConnections[i].Read(buffer)
+		require.Equal(t, 0, read, "local transport should disconnect from test peer without reading anything")
+		require.Error(t, err, "local transport should disconnect from test peer")
+	}
+}
+
+func TestDirectOutgoing_ConnectionReconnectsOnFailure(t *testing.T) {
+	t.Parallel()
+	test.WithContext(func(ctx context.Context) {
+
+		h := newDirectHarness().start(ctx)
+
+		listener, err := net.Listen("tcp", fmt.Sprintf(":%d", h.portForPeer(0)))
+		defer listener.Close()
+		require.NoError(t, err, "test peer server could not listen")
+
+		connection, err := listener.Accept()
+		defer connection.Close()
```

Comment:
I'm not sure about this defer. Does it close the connection pointed from `connection` or the specific instance?  if so, then the connection it points to is closed in the first forceful disconnect...

Reply:
There's already a separate issue for this problem:
https://github.com/orbs-network/orbs-network-go/issues/187

I added your issue there as well.

We should resolve this separately because I invested significant efforts to make sure all resources are closed yet this still happens.. require further investigation and I don't want to block this PR